File: src/main/java/org/stevenlowes/tools/lifx_controller/Messages/Device/StateInfo.java
package org.stevenlowes.tools.lifx_controller.Messages.Device;

import org.stevenlowes.tools.lifx_controller.Messages.DataTypes.Payload;

import java.math.BigInteger;

public class StateInfo extends Payload {
    private final int code = 35;
    private BigInteger time;                // 64-Bits (Unsigned)
    private BigInteger uptime;                // 64-Bits (Unsigned)
    private BigInteger downtime;            // 64-Bits (Unsigned)

    public StateInfo() {
        time = BigInteger.valueOf(0L);
        uptime = BigInteger.valueOf(0L);
        downtime = BigInteger.valueOf(0L);
    }

    public StateInfo(BigInteger time, BigInteger uptime, BigInteger downtime) {
        this.time = time;
        this.uptime = uptime;
        this.downtime = downtime;
    }

    public StateInfo(StateInfo stateInfo) {
        time = stateInfo.time;
        uptime = stateInfo.uptime;
        downtime = stateInfo.downtime;
    }

    public int getCode() {
        return code;
    }

    public BigInteger getTime() {
        return time;
    }

    public void setTime(BigInteger time) {
        this.time = time;
    }

    public BigInteger getUptime() {
        return uptime;
    }

    public void setUptime(BigInteger uptime) {
        this.uptime = uptime;
    }

    public BigInteger getDowntime() {
        return downtime;
    }

    public void setDowntime(BigInteger downtime) {
        this.downtime = downtime;
    }

    public void setFromCommandByteArray(byte[] byteArray) {
        byte[] timeBytes = new byte[8];
        for (int i = 43; i > 35; i--) {
            timeBytes[(-1 * i) + 43] = byteArray[i];
        }
        time = new BigInteger(timeBytes);

        byte[] uptimeBytes = new byte[8];
        for (int i = 51; i > 43; i--) {
            uptimeBytes[(-1 * i) + 51] = byteArray[i];
        }
        uptime = new BigInteger(uptimeBytes);

        byte[] downtimeBytes = new byte[8];
        for (int i = 59; i > 51; i--) {
            downtimeBytes[(-1 * i) + 59] = byteArray[i];
        }
        downtime = new BigInteger(downtimeBytes);

    }

}

File: src/main/java/org/stevenlowes/tools/lifxcontroller/messages/device/StateInfo.kt
package org.stevenlowes.tools.lifxcontroller.messages.device

import org.stevenlowes.tools.lifxcontroller.messages.datatypes.SetOnlyPayload

import java.math.BigInteger

class StateInfo(var time: BigInteger = BigInteger.ZERO, var uptime: BigInteger = BigInteger.ZERO, var downtime: BigInteger = BigInteger.ZERO) : SetOnlyPayload(35) {

    constructor(stateInfo: StateInfo): this(stateInfo.time, stateInfo.uptime, stateInfo.downtime)

    override fun setFromCommandByteArray(byteArray: ByteArray) {
        val timeBytes = ByteArray(8)
        for (i in 43 downTo 36) {
            timeBytes[-1 * i + 43] = byteArray[i]
        }
        time = BigInteger(timeBytes)

        val uptimeBytes = ByteArray(8)
        for (i in 51 downTo 44) {
            uptimeBytes[-1 * i + 51] = byteArray[i]
        }
        uptime = BigInteger(uptimeBytes)

        val downtimeBytes = ByteArray(8)
        for (i in 59 downTo 52) {
            downtimeBytes[-1 * i + 59] = byteArray[i]
        }
        downtime = BigInteger(downtimeBytes)

    }

}
